install.sh

```diff
@@ -0,0 +1,156 @@
+#!/bin/bash
+# File: install.bash
+#
+# Author: Jayden Leong <jdleong58@gmail.com>
+#
+# Date: January 31, 2021
+#
+# Purpose: install piponic project on Raspberry Pi Zero W V1.1
+
+SHOW_HELP() {
+  echo
+  "Usage:
+    sudo ./setup.sh <DEVICE-ID> <REGISTRY-ID> <PROJECT-ID>
+
+  Example:
+    sudo ./setup.sh JaydenPi RaspberryPis piponics
+    
+  Arguments:
+    <DEVICE-ID> : Google Cloud Device ID of this Raspberry Pi
+    <REGISTRY-ID> : Google Cloud Registry this device is part of
+    <PROJECT-ID> : Google Cloud Project name
+   
+    Please consult the Google Cloud Console -> IoT Core to create
+    a new device before running this script."
+}
+
+# Ensure correct arguments
+if [ "$#" -ne 3 ]; then
+    echo "Incorrect number of parameters"
+    SHOW_HELP
+    exit 1
+fi
+
+# Parse arguments
+DEVICE_ID=$1
+REGISTRY_ID=$2
+PROJECT_ID=$3
+
+# Get directory this script is in
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+echo ""
+echo "----- INSTALLING APT PACKAGES -----"
+echo ""
+
+apt-get update
+apt-get install -yy vim git python3 python3-pip python3-rpi.gpio
+
+echo ""
+echo "----- INSTALLING PYTHON DEPENDENCIES FROM REQUIREMENTS.TXT -----"
+echo ""
+pip3 install -r requirements.txt
+
+if [ ! -f ${SCRIPT_DIR}/rsa_private.pem ]; then
+    echo ""
+    echo "----- GENERATING RSA KEYS FOR GOOGLE CLOUD COMMUNICATION -----"
+    echo ""
+    openssl genpkey -algorithm RSA -out rsa_private.pem -pkeyopt rsa_keygen_bits:2048
+    openssl rsa -in rsa_private.pem -pubout -out rsa_public.pem
+    chmod +r rsa_private.pem rsa_public.pem
+fi 
+
+# Download google root CA
+# See here for details https://cloud.google.com/iot/docs/how-tos/mqtt-bridge
+if [ ! -f ${SCRIPT_DIR}/roots.pem ]; then
+    echo ""
+    echo "----- DOWNLOADING GOOGLE ROOT RA -----"
+    echo ""
+    curl https://pki.goog/roots.pem --output roots.pem
+fi
+
+if ! command -v gcloud &> /dev/null
+then 
+    echo ""
+    echo "----- INSTALLING GOOGLE CLOUD SDK -----"
+    echo ""
+    
+    echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] https://packages.cloud.google.com/apt cloud-sdk main" | sudo tee -a /etc/apt/sources.list.d/google-cloud-sdk.list
+    apt-get install -yy apt-transport-https ca-certificates gnupg
+    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key --keyring /usr/share/keyrings/cloud.google.gpg add -
+    apt-get install -yy google-cloud-sdk
+    gcloud init --console-only
+else 
+    gcloud auth login
+    gcloud config set project ${PROJECT_ID}
+fi 
+
+# Create IoT Device registry if not exists
+if gcloud iot registries list --project ${PROJECT_ID} --region us-central1 | grep -q $REGISTRY_ID
```

Comment:
Are there any implications for doing this if you misspell your project_ID? I assume it won't connect with the correct firebase app, but other ones, like incurred cost/billing from Google?

Reply:
Good point, it will fail to setup the device properly. I will try to put an error check in the script.

Reply:
Now addressed. See `if` statement above.

Reply:
Nice, looks good. I'll test it today during our labtime.